File: wordpress-plugin/animeh/src/Support/AssScript.php
<?php
/**
 * ASS/SSA script reading.
 *
 * The PHP counterpart of the player's `src/subtitles/ass.ts`. Only enough
 * structure is read to know which fonts a script needs — rendering is libass's
 * job, in the browser.
 *
 * The two implementations are held to the same expected output by the test
 * suites on both sides, so a change to one that the other misses shows up
 * immediately.
 *
 * @package Animeh
 */

declare( strict_types = 1 );

namespace Animeh\Support;

/**
 * Reads font requirements out of an ASS script.
 */
final class AssScript {

	/**
	 * Every font family a script can ask for.
	 *
	 * Two places name fonts: the `Fontname` column of each style, and `\fn`
	 * override tags inside dialogue text. Missing either one means a line
	 * renders in the wrong face at playback time, which is exactly the failure
	 * the font report exists to catch.
	 *
	 * @param string $content Script contents.
	 * @return string[] Sorted, de-duplicated family names.
	 */
	public static function font_families( string $content ): array {
		$families = array();

		foreach ( self::styles( $content ) as $style ) {
			if ( '' !== $style['fontname'] ) {
				$families[ FontFile::key( $style['fontname'] ) ] = $style['fontname'];
			}
		}

		// `\fn` runs until the next backslash or the end of the override block.
		if ( preg_match_all( '/\\\\fn([^\\\\}]*)/u', $content, $matches ) ) {
			foreach ( $matches[1] as $raw ) {
				$name = self::normalise_font_name( $raw );
				if ( '' !== $name ) {
					$families[ FontFile::key( $name ) ] = $name;
				}
			}
		}

		$values = array_values( $families );
		usort( $values, static fn( string $a, string $b ): int => strcasecmp( $a, $b ) );
		return $values;
	}

	/**
	 * Styles declared by the script.
	 *
	 * @param string $content Script contents.
	 * @return array<int, array{name: string, fontname: string, fontsize: float, bold: bool, italic: bool}>
	 */
	public static function styles( string $content ): array {
		$section = self::section( $content, array( 'v4+ styles', 'v4 styles' ) );
		if ( array() === $section ) {
			return array();
		}

		$format = array();
		$styles = array();

		foreach ( $section as $line ) {
			$entry = self::split_entry( $line );
			if ( null === $entry ) {
				continue;
			}
			$key = strtolower( $entry[0] );

			if ( 'format' === $key ) {
				$format = array_map(
					static fn( string $field ): string => strtolower( trim( $field ) ),
					explode( ',', $entry[1] )
				);
				continue;
			}

			if ( 'style' !== $key || array() === $format ) {
				continue;
			}

			// Style values are comma-separated and positional.
			$fields = explode( ',', $entry[1] );
			$get    = static function ( string $name ) use ( $format, $fields ): string {
				$index = array_search( $name, $format, true );
				return false === $index ? '' : trim( $fields[ $index ] ?? '' );
			};

			$name = $get( 'name' );
			if ( '' === $name ) {
				continue;
			}

			$bold   = $get( 'bold' );
			$italic = $get( 'italic' );

			$styles[] = array(
				'name'     => $name,
				'fontname' => self::normalise_font_name( $get( 'fontname' ) ),
				'fontsize' => (float) $get( 'fontsize' ),
				// ASS booleans are -1 for true, 0 for false.
				'bold'     => ( '-1' === $bold || '1' === $bold ),
				'italic'   => ( '-1' === $italic || '1' === $italic ),
			);
		}

		return $styles;
	}

	/**
	 * Script resolution, when declared.
	 *
	 * @param string $content Script contents.
	 * @return array{x: int|null, y: int|null}
	 */
	public static function play_res( string $content ): array {
		$result = array(
			'x' => null,
			'y' => null,
		);
		foreach ( self::section( $content, array( 'script info' ) ) as $line ) {
			if ( str_starts_with( ltrim( $line ), ';' ) ) {
				continue;
			}
			$entry = self::split_entry( $line );
			if ( null === $entry ) {
				continue;
			}
			$key = strtolower( $entry[0] );
			if ( 'playresx' === $key ) {
				$result['x'] = (int) $entry[1];
			} elseif ( 'playresy' === $key ) {
				$result['y'] = (int) $entry[1];
			}
		}
		return $result;
	}

	/**
	 * Number of dialogue events.
	 *
	 * @param string $content Script contents.
	 */
	public static function dialogue_count( string $content ): int {
		$count = 0;
		foreach ( self::section( $content, array( 'events' ) ) as $line ) {
			$entry = self::split_entry( $line );
			if ( null !== $entry && 'dialogue' === strtolower( $entry[0] ) ) {
				++$count;
			}
		}
		return $count;
	}

	/**
	 * Strip the decorations ASS allows around a font name.
	 *
	 * A leading `@` requests the vertical-writing variant of a CJK face; the
	 * family being asked for is the same one either way.
	 *
	 * @param string $name Raw name.
	 */
	public static function normalise_font_name( string $name ): string {
		return trim( ltrim( trim( $name ), '@' ) );
	}

	/**
	 * Lines belonging to the first matching section.
	 *
	 * @param string   $content Script contents.
	 * @param string[] $names   Section names to try, lowercased, in order.
	 * @return string[]
	 */
	private static function section( string $content, array $names ): array {
		// Strip a BOM: libass copes, but our own field parsing would not.
		$content  = preg_replace( '/^\xEF\xBB\xBF/', '', $content ) ?? $content;
		$sections = array();
		$current  = null;

		foreach ( preg_split( '/\r\n|\r|\n/', $content ) ?: array() as $line ) {
			if ( preg_match( '/^\s*\[([^\]]+)\]\s*$/', $line, $matches ) ) {
				$current              = strtolower( trim( $matches[1] ) );
				$sections[ $current ] = array();
				continue;
			}
			if ( null !== $current ) {
				$sections[ $current ][] = $line;
			}
		}

		foreach ( $names as $name ) {
			if ( isset( $sections[ $name ] ) ) {
				return $sections[ $name ];
			}
		}
		return array();
	}

	/**
	 * Split a `Key: value` line.
	 *
	 * @param string $line Line to split.
	 * @return array{0: string, 1: string}|null
	 */
	private static function split_entry( string $line ): ?array {
		$index = strpos( $line, ':' );
		if ( false === $index ) {
			return null;
		}
		return array( trim( substr( $line, 0, $index ) ), trim( substr( $line, $index + 1 ) ) );
	}
}

File: android/app/src/main/java/com/animeh/app/player/ass/AssScript.kt
package com.animeh.app.player.ass

/**
 * A subtitle script, parsed far enough to draw it.
 *
 * This exists because going through ExoPlayer's text track threw away the two
 * things that matter most for ASS. A `Cue` is the rendered text and a place to
 * put it: the **style is resolved away**, so there is no way to know that the
 * sign at the top of the frame and the dialogue at the bottom are meant to be
 * different fonts, sizes and colours. And the track has to be selected,
 * loaded and decoded by the player before a single cue arrives, which is a
 * chain of things that can quietly not happen — an episode resumed part-way
 * through, a track the selector did not pick, a lazily loaded source nobody
 * asked for. Subtitles that need the track turning off and on again to appear
 * are that chain failing.
 *
 * The script is already downloaded, to find out which fonts it needs. Reading
 * the rest of it costs nothing and answers both: every line carries its own
 * style, and which lines are on screen is a comparison against the playhead
 * rather than something to be waited for.
 *
 * ## What this reads, and what it does not
 *
 * Read: script resolution, styles (font, size, weight, slant, colours,
 * outline, alignment, margins), dialogue timing, and the inline overrides that
 * change any of those for one line — `\an`, `\pos`, `\fn`, `\fs`, `\b`, `\i`,
 * `\c`, `\1c`, `\3c`, `\bord`.
 *
 * Not read: karaoke timing (`\k`), transforms (`\t`), animation (`\move`,
 * `\fad`), rotation (`\frx`, `\fry`, `\frz`), and vector drawing (`\p`). The
 * first four degrade to the plain line, which is the honest failure. A drawing
 * is skipped outright rather than printed — a line of `m 132 72 l 423 72` on
 * screen would be worse than the shape being absent.
 *
 * An override applies to the whole line rather than from where it appears.
 * Scripts that change colour mid-sentence therefore take the first colour;
 * doing better needs per-character runs, which is the next honest step and not
 * this one.
 */
data class AssStyle(
    val name: String,
    val family: String,
    val sizePx: Float,
    val bold: Boolean,
    val italic: Boolean,
    val fill: Int,
    val outline: Int,
    val outlineWidth: Float,
    val alignment: Int,
    val marginL: Int,
    val marginR: Int,
    val marginV: Int,
)

/** One line of dialogue, with every style question already answered. */
data class AssLine(
    val startMs: Long,
    val endMs: Long,
    /** Draw order: lower layers first, so a sign's shadow sits under it. */
    val layer: Int,
    /** The text as it should appear; `\N` is already a newline. */
    val text: String,
    val family: String,
    /** In script units — see [AssScript.playResY]. */
    val sizePx: Float,
    val bold: Boolean,
    val italic: Boolean,
    val fill: Int,
    val outline: Int,
    val outlineWidth: Float,
    /** Numpad alignment, 1 to 9. */
    val alignment: Int,
    val marginL: Int,
    val marginR: Int,
    val marginV: Int,
    /** From `\pos`, in script units, or null when the line is not placed. */
    val posX: Float?,
    val posY: Float?,
)

/**
 * A parsed script.
 *
 * [playResX] and [playResY] are the coordinate system every measurement in
 * [lines] is expressed in; the renderer scales from them to the surface it has.
 */
class AssScript(
    val playResX: Int,
    val playResY: Int,
    val lines: List<AssLine>,
    /** The family the dialogue style uses, for lines that ask for nothing. */
    val primaryFamily: String?,
) {

    /** Every family the script asks for, in the order first seen. */
    val families: List<String> = lines.map { it.family }.distinct()

    /**
     * The lines on screen at this moment, in the order they should be drawn.
     *
     * A linear scan: a script is a few hundred lines and this runs ten times a
     * second, which is nothing next to decoding a frame of video.
     */
    fun activeAt(positionMs: Long): List<AssLine> =
        lines.filter { positionMs >= it.startMs && positionMs < it.endMs }
            .sortedBy { it.layer }

    val isEmpty: Boolean get() = lines.isEmpty()
}

/**
 * Reading a script into [AssScript].
 *
 * Kept separate from [AssParser] — which answers "which fonts does this need"
 * without any notion of drawing — because this is the part that has to be
 * right about numbers, and it is worth being able to test on its own.
 */
object AssReader {

    /**
     * Parse a whole script.
     *
     * Never throws. A field it cannot read falls back to the style's value,
     * and a style it cannot read falls back to a plain white sans — a subtitle
     * in the wrong font is a subtitle, and an exception is a black screen.
     */
    fun read(content: String): AssScript {
        var playResX = 384
        var playResY = 288

        val styles = LinkedHashMap<String, AssStyle>()
        val lines = mutableListOf<AssLine>()

        var section = ""
        var styleFormat: List<String> = emptyList()
        var eventFormat: List<String> = emptyList()

        content.lineSequence().forEach { rawLine ->
            val line = rawLine.trim()

            if (line.startsWith("[") && line.endsWith("]")) {
                section = line.trim('[', ']').lowercase()
                return@forEach
            }

            when {
                section.startsWith("script") -> when {
                    line.startsWith("PlayResX:", true) ->
                        line.removePrefix("PlayResX:").trim().toIntOrNull()?.let { playResX = it }
                    line.startsWith("PlayResY:", true) ->
                        line.removePrefix("PlayResY:").trim().toIntOrNull()?.let { playResY = it }
                }

                section.contains("styles") -> when {
                    line.startsWith("Format:", true) ->
                        styleFormat = columns(line.removePrefix("Format:"))

                    line.startsWith("Style:", true) ->
                        style(line.removePrefix("Style:"), styleFormat)?.let { styles[it.name.lowercase()] = it }
                }

                section.contains("events") -> when {
                    line.startsWith("Format:", true) ->
                        eventFormat = columns(line.removePrefix("Format:"))

                    // `Comment:` lines are the script's own notes and are not
                    // drawn; this file uses them to mark the ending and the
                    // preview, which would otherwise appear as dialogue.
                    line.startsWith("Dialogue:", true) ->
                        event(line.removePrefix("Dialogue:"), eventFormat, styles)?.let { lines += it }
                }
            }
        }

        val primary = styles["default"]?.family ?: styles.values.firstOrNull()?.family

        return AssScript(
            playResX = playResX.coerceAtLeast(1),
            playResY = playResY.coerceAtLeast(1),
            lines = lines.sortedBy { it.startMs },
            primaryFamily = primary,
        )
    }

    /** One `Style:` row. */
    private fun style(rest: String, format: List<String>): AssStyle? {
        val fields = rest.split(",").map { it.trim() }

        fun at(name: String, fallback: Int): String =
            fields.getOrNull(format.indexOf(name).takeIf { it >= 0 } ?: fallback).orEmpty()

        val name = at("name", 0)
        val family = AssParser.normalise(at("fontname", 1)) ?: return null

        if (name.isBlank()) return null

        return AssStyle(
            name = name,
            family = family,
            sizePx = at("fontsize", 2).toFloatOrNull() ?: 48f,
            // ASS booleans are 0 and -1, and a few tools write 1.
            bold = at("bold", 7).trim().let { it == "-1" || it == "1" },
            italic = at("italic", 8).trim().let { it == "-1" || it == "1" },
            fill = colour(at("primarycolour", 3)) ?: WHITE,
            outline = colour(at("outlinecolour", 5)) ?: BLACK,
            outlineWidth = at("outline", 16).toFloatOrNull() ?: 2f,
            alignment = at("alignment", 18).toIntOrNull()?.takeIf { it in 1..9 } ?: 2,
            marginL = at("marginl", 19).toIntOrNull() ?: 0,
            marginR = at("marginr", 20).toIntOrNull() ?: 0,
            marginV = at("marginv", 21).toIntOrNull() ?: 0,
        )
    }

    /** One `Dialogue:` row, with its style applied and its overrides on top. */
    private fun event(
        rest: String,
        format: List<String>,
        styles: Map<String, AssStyle>,
    ): AssLine? {
        val textIndex = format.indexOf("text").takeIf { it >= 0 } ?: 9

        // Only the fields before the text are comma-separated; the text itself
        // may contain commas and is everything left over.
        val fields = rest.split(",", limit = textIndex + 1)
        if (fields.size <= textIndex) return null

        fun at(name: String, fallback: Int): String =
            fields.getOrNull(format.indexOf(name).takeIf { it >= 0 } ?: fallback).orEmpty().trim()

        val startMs = time(at("start", 1)) ?: return null
        val endMs = time(at("end", 2)) ?: return null
        if (endMs <= startMs) return null

        val raw = fields[textIndex]

        // A drawing is a path, not words. Printing its commands would put
        // "m 132 72 l 423 72" on screen, which is worse than the shape being
        // missing.
        if (DRAWING.containsMatchIn(raw)) return null

        val style = styles[at("style", 3).lowercase()]
            ?: styles.values.firstOrNull()
            ?: FALLBACK_STYLE

        val text = plain(raw)
        if (text.isBlank()) return null

        val overrides = raw.let { OVERRIDE_BLOCK.findAll(it).joinToString("") { m -> m.value } }

        val position = POSITION.find(overrides)
        val alignment = (ALIGNMENT.find(overrides)?.groupValues?.get(1)
            ?: LEGACY_ALIGNMENT.find(overrides)?.groupValues?.get(1))
            ?.toIntOrNull()
            ?.takeIf { it in 1..9 }

        return AssLine(
            startMs = startMs,
            endMs = endMs,
            layer = at("layer", 0).toIntOrNull() ?: 0,
            text = text,
            family = FAMILY.find(overrides)?.groupValues?.get(1)?.let { AssParser.normalise(it) } ?: style.family,
            sizePx = SIZE.find(overrides)?.groupValues?.get(1)?.toFloatOrNull() ?: style.sizePx,
            bold = BOLD.find(overrides)?.groupValues?.get(1)?.let { it != "0" } ?: style.bold,
            italic = ITALIC.find(overrides)?.groupValues?.get(1)?.let { it != "0" } ?: style.italic,
            fill = FILL.find(overrides)?.groupValues?.get(1)?.let { colour(it) } ?: style.fill,
            outline = BORDER_COLOUR.find(overrides)?.groupValues?.get(1)?.let { colour(it) } ?: style.outline,
            outlineWidth = BORDER.find(overrides)?.groupValues?.get(1)?.toFloatOrNull() ?: style.outlineWidth,
            alignment = alignment ?: style.alignment,
            // A line's own margin of zero means "use the style's".
            marginL = at("marginl", 5).toIntOrNull()?.takeIf { it > 0 } ?: style.marginL,
            marginR = at("marginr", 6).toIntOrNull()?.takeIf { it > 0 } ?: style.marginR,
            marginV = at("marginv", 7).toIntOrNull()?.takeIf { it > 0 } ?: style.marginV,
            posX = position?.groupValues?.get(1)?.trim()?.toFloatOrNull(),
            posY = position?.groupValues?.get(2)?.trim()?.toFloatOrNull(),
        )
    }

    /** `H:MM:SS.cc` as milliseconds. */
    fun time(value: String): Long? {
        val parts = value.trim().split(":")
        if (parts.size != 3) return null

        val hours = parts[0].toLongOrNull() ?: return null
        val minutes = parts[1].toLongOrNull() ?: return null
        val seconds = parts[2].toDoubleOrNull() ?: return null

        return ((hours * 3600 + minutes * 60) * 1000) + (seconds * 1000).toLong()
    }

    /**
     * `&HAABBGGRR` as ARGB.
     *
     * Two things catch people out: the channels are backwards, and the first
     * byte is *transparency* rather than opacity — `&H00FFFFFF` is opaque
     * white, not invisible white.
     */
    fun colour(value: String): Int? {
        val hex = value.trim().removePrefix("&H").removePrefix("&h").removeSuffix("&")
            .takeWhile { it.isDigit() || it in 'a'..'f' || it in 'A'..'F' }

        if (hex.isEmpty() || hex.length > 8) return null

        val packed = hex.toLongOrNull(16) ?: return null

        val blue = ((packed shr 16) and 0xFF).toInt()
        val green = ((packed shr 8) and 0xFF).toInt()
        val red = (packed and 0xFF).toInt()
        val transparency = if (hex.length > 6) ((packed shr 24) and 0xFF).toInt() else 0

        return ((255 - transparency) shl 24) or (red shl 16) or (green shl 8) or blue
    }

    /** The text as it should read: overrides gone, breaks real. */
    fun plain(raw: String): String =
        raw.replace(OVERRIDE_BLOCK, "")
            .replace(LINE_BREAK, "\n")
            .replace(HARD_SPACE, " ")
            .trim()

    private fun columns(rest: String): List<String> =
        rest.split(",").map { it.trim().lowercase() }

    private const val WHITE = 0xFFFFFFFF.toInt()
    private const val BLACK = 0xFF000000.toInt()

    private val FALLBACK_STYLE = AssStyle(
        name = "Default",
        family = "sans-serif",
        sizePx = 48f,
        bold = false,
        italic = false,
        fill = WHITE,
        outline = BLACK,
        outlineWidth = 2f,
        alignment = 2,
        marginL = 0,
        marginR = 0,
        marginV = 0,
    )

    private val OVERRIDE_BLOCK = Regex("""\{[^}]*\}""")
    private val LINE_BREAK = Regex("""\\[Nn]""")
    private val HARD_SPACE = Regex("""\\h""")

    /** A non-zero `\p` turns the line into vector drawing commands. */
    private val DRAWING = Regex("""\\p[1-9]""")

    private val POSITION = Regex("""\\pos\(([^,]+),([^)]+)\)""")
    private val ALIGNMENT = Regex("""\\an(\d)""")
    private val LEGACY_ALIGNMENT = Regex("""\\a(\d{1,2})(?!n)""")
    private val FAMILY = Regex("""\\fn([^\\}]+)""")
    private val SIZE = Regex("""\\fs([\d.]+)""")
    private val BOLD = Regex("""\\b([01])(?![a-z])""")
    private val ITALIC = Regex("""\\i([01])(?![a-z])""")
    private val FILL = Regex("""\\1?c(&H[0-9a-fA-F]+&?)""")
    private val BORDER_COLOUR = Regex("""\\3c(&H[0-9a-fA-F]+&?)""")
    private val BORDER = Regex("""\\bord([\d.]+)""")
}
